go/login/server/keymanager.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"../../glome"
+)
+
+// ErrInvalidKeyIndex denotes that provided index is invalid for a key. Only
+// indexes in range 0-127 are valid.
+type ErrInvalidKeyIndex struct {
+	Index uint8
+}
+
+func (e ErrInvalidKeyIndex) Error() string {
+	return fmt.Sprintf("key index should be in range 0-127, found: %v", e.Index)
+}
+
+// ErrOverloadedKeyIndex denotes that provided index is already in use, so no
+// new keys can be assigned to it.
+type ErrOverloadedKeyIndex struct {
+	Index uint8
+}
+
+func (e ErrOverloadedKeyIndex) Error() string {
+	return fmt.Sprintf("key index already in use, found: %v", e.Index)
+}
+
+// ErrKeyIndexNotFound denotes that an index was not found, so no key can be
+// returned from it.
+type ErrKeyIndexNotFound struct {
+	Index uint8
+}
+
+func (e ErrKeyIndexNotFound) Error() string {
+	return fmt.Sprintf("key index %v not found", e.Index)
+}
+
+// A PrivateKey represent a Private key for the login server. It is a pair composed of a private key
+// and its pairing index.
+type PrivateKey struct {
+	Value glome.PrivateKey
+	Index uint8
+}
+
+// A PublicKey represent a Service key for the login server. It is a pair composed of a public key
+// and its pairing index.
+type PublicKey struct {
+	Value glome.PublicKey
+	Index uint8
+}
+
+// KeyManager performs key maneger task in a concurrent-safe way. It allows for constant
+// time search of keys by index.
+type KeyManager struct {
+	indexToPriv map[uint8]glome.PrivateKey
+	publicKeys  []PublicKey
+	lock        sync.RWMutex
+}
+
+func (k *KeyManager) asyncAdd(key glome.PrivateKey, index uint8) error {
+	if index > 127 { // Maybe having a constant in the login library to store this (?)
```

Comment:
Since the login library does not define such a constant, consider removing this comment.

Reply:
Deleted.